feat: close_dlc msg and cooperative close logic

matthewjablack opened the pull request:
## What

Add `close_dlc` msg type, and cooperative close logic

## Why

It allows Alice or Bob to pass a message with the intention of cooperatively closing the DLC, allowing for early exit from a position.

The initiator provides an input, so they can cancel if the counterparty doesn't respond right away.

Yes this creates a free option problem, but the expectation is that the counterparty is a trusted party (for example a market maker). 

This message type was discussed in DLC Specs: https://github.com/discreetlogcontracts/dlcspecs/issues/161

And has been implemented in the Atomic Finance typescript libraries: https://github.com/AtomicFinance/bitcoin-abstraction-layer/pull/86 and https://github.com/AtomicFinance/node-dlc/pull/122

matthewjablack committed:
feat: close_dlc msg and cooperative close logic

Tibo-lg commented on dlc-manager/src/manager.rs:
```diff
@@ -470,6 +474,47 @@ where
+            ),
+        };
+
+        self.store.update_contract(&Contract::Closed(closed_contract))?;
```
Would be better to use `PreClosed` here. And so set `signed_cet` to the closing tx.

matthewjablack replied:
Good catch, fixed in [22c22c6](https://github.com/p2pderivatives/rust-dlc/pull/255/commits/22c22c62d2f653ed68840f14d26ca958d783ac62)

Tibo-lg commented on dlc-manager/src/contract_updater.rs:
```diff
@@ -754,6 +755,129 @@ where
+}
+
+/// Verifies and completes a cooperative close transaction using the counter party's signature.
+pub fn verify_and_complete_cooperative_close<C: Signing, SP: Deref>(
```
nit: I don't think it's verifying anything here so maybe you can drop the `verify_`

matthewjablack replied:
Good point, fixed in [22c22c6](https://github.com/p2pderivatives/rust-dlc/pull/255/commits/22c22c62d2f653ed68840f14d26ca958d783ac62)

Tibo-lg commented on dlc-manager/src/contract/mod.rs:
```diff
@@ -43,6 +43,8 @@ pub enum Contract {
     /// A contract whose refund transaction was broadcast.
     Refunded(signed_contract::SignedContract),
+    /// A contract that was closed cooperatively.
+    CooperativeClose(CooperativeCloseContract),
```
What is the point of that state? I feel the best would be to just not updated the state when proposing, and then update if/when the closing transaction is seen on-chain. Right now if you propose a cooperative close but it somehow doesn't end up being closed you will be in an incorrect state.

matthewjablack replied:
Agreed, removed this state in [22c22c6](https://github.com/p2pderivatives/rust-dlc/pull/255/commits/22c22c62d2f653ed68840f14d26ca958d783ac62)

Tibo-lg commented on dlc-manager/src/manager.rs:
```diff
@@ -470,6 +474,47 @@ where
+        let signed_contract = get_contract_in_state!(
+            self,
+            &close_msg.contract_id,
+            Signed,
```
Should be `Confirmed`

matthewjablack replied:
Good catch, fixed in [22c22c6](https://github.com/p2pderivatives/rust-dlc/pull/255/commits/22c22c62d2f653ed68840f14d26ca958d783ac62)

Tibo-lg reviewed (changes requested):
It's a good feature, though:
* please add an integration test so we can be confident it's working, I don't think it does right now
* The chain monitor is used for channels, I don't think what you implemented here will work
* I think the best way to go to close the contract on the party offering the cooperative close would be to stay in the `Confirmed` state, but then maybe have a `close_tx: Option<Transaction>` field that you check within `check_confirmed_contract` so that you can mark the contract as `PreClosed/Closed`. It could also be a list if we assume that there could be multiple cooperative close offers (in case some are not accepted/processed).

matthewjablack committed:
fix: cooperative close implementation issues

- rm problematic CooperativeClose state that could persist indefinitely
- fix state validation to expect Confirmed instead of Signed state
- rm chain monitor usage (designed for channels, not standalone contracts)
- add integration tests
- rename verify_and_complete_cooperative_close to complete_cooperative_close
- update storage and serialization to remove CooperativeClose variant

Fixes issues identified in code review where CooperativeClose state
could get stuck if counterparty never responds, and resolves transaction
broadcast conflicts with chain monitor architecture.

matthewjablack commented:
Thanks for that review @Tibo-lg 

> please add an integration test so we can be confident it's working, I don't think it does right now

Added to `manager_execution_tests.rs`

> The chain monitor is used for channels, I don't think what you implemented here will work

Thanks for clarification, removed accordingly

> I think the best way to go to close the contract on the party offering the cooperative close would be to stay in the Confirmed state, but then maybe have a close_tx: Option<Transaction> field that you check within check_confirmed_contract so that you can mark the contract as PreClosed/Closed. It could also be a list if we assume that there could be multiple cooperative close offers (in case some are not accepted/processed).

Not sure what you had in mind here. Were you thinking of having `close_tx: Option<Transaction>` be added to an existing struct? Which one did you have in mind? 

I'm guessing you were thinking of adding it to `DlcTransactions`

matthewjablack committed:
lint: apply rustfmt formatting fixes

matthewjablack committed:
lint: clippy manual_ok_err warning in hex_utils

- replace manual Ok/Err match with .ok() method
- fixes CI failure on clippy:manual_ok_err lint

Tibo-lg commented on dlc-manager/src/manager.rs:
```diff
@@ -470,6 +474,65 @@ where
+        };
+
+        self.store
+            .update_contract(&Contract::Closed(closed_contract))?;
```
PreClosed

matthewjablack replied:
Thanks, fixed in 3356604

Tibo-lg commented on dlc-manager/tests/manager_execution_tests.rs:
```diff
@@ -806,93 +848,194 @@ fn manager_execution_test(test_params: TestParams, path: TestPath, manual_close:
+                        .expect("Error accepting cooperative close");
+
+                    // Bob should now be in Closed state (he broadcast the transaction)
+                    assert_contract_state!(bob_manager_send, contract_id, Closed);
```
Should be PreClosed because the transaction is not confirmed.

matthewjablack replied:
Good catch, fixed in 3356604

Tibo-lg commented on dlc-manager/tests/manager_execution_tests.rs:
```diff
@@ -806,93 +848,194 @@ fn manager_execution_test(test_params: TestParams, path: TestPath, manual_close:
+                    // Mine a block to confirm the close transaction
+                    generate_blocks(1);
+
+                    // In a real scenario, Alice would detect the close transaction and call on_counterparty_close
+                    // For the test, we'll verify the cooperative close functionality worked correctly
```
We definitely want to detect the close transaction, and we want to test it.

matthewjablack replied:
I've added `check_pending_close_transactions` function and included it in `periodic_check` https://github.com/p2pderivatives/rust-dlc/pull/255/files#diff-6901f71d060cd9a08d68e0e45ba8e40bcd714a041ddcdff6ea168eb4b35a1c4dR813-R863

Also updated tests https://github.com/p2pderivatives/rust-dlc/pull/255/files#diff-11a961d721d78a3c54a7b2a0b90b92b5a3768c0fbed6fef9c45014c0be8702e8R910-R936

Tibo-lg commented on dlc-manager/src/manager.rs:
```diff
@@ -470,6 +474,65 @@ where
         Ok(())
     }
 
+    fn on_close_message(
```
I'm not sure we should have that to be honest. It feels like having only the `accept` method is better. This means I can just send a close message and have it accepted by default.

matthewjablack replied:
Agreed, I've updated it to simply validate the close message

Tibo-lg replied:
Is the idea that an application should implement a custom message handler to accept it?

matthewjablack replied:
Yep exactly. When closing a contract, the acceptor should double check that the payout matches their expected value of the contract, or lines up with their own business logic for them being comfortable with closing the contract early.

Tibo-lg commented:
> Not sure what you had in mind here. Were you thinking of having close_tx: Option<Transaction> be added to an existing struct? Which one did you have in mind?

> I'm guessing you were thinking of adding it to DlcTransactions

I don't have a strong opinion on where to put it, I guess `DlcTransactions` struct is fine.

matthewjablack committed:
feat: add automatic cooperative close detection

- add `pending_close_txs` field to track cooperative close offers
- add `check_pending_close_transactions()` for chain monitoring
- fix cooperative close flow: validate vs accept separation
- implement proper state transitions: Confirmed → PreClosed → Closed
- update test coverage for cooperative close detection
- support self-initiated and counterparty-initiated closes

Resolves automatic detection of cooperative close transactions that are
broadcast by counterparties, ensuring both parties transition to the
correct final state

matthewjablack committed:
lint: clippy inneficient swaps w/ assignment

matthewjablack committed:
test: update serialized files for pending_close_txs

Regenerated test files after adding pending_close_txs field to
Dlctransactions. These serialized files were used by the sled storage
provider tests to verify correct serialization/deserialization.

matthewjablack commented:
Thanks @Tibo-lg 

I added `pending_close_txs` field to `DlcTransactions` to track cooperative close offers, `check_pending_close_transactions()` for chain. monitoring

Also fixed proper state transitions, with Confirmed → PreClosed → Closed

bennyhodl commented on dlc-manager/src/contract_updater.rs:
```diff
@@ -754,6 +758,127 @@ where
+    );
+
+    // Get our private key and sign the transaction
+    let fund_private_key =
```
Instead of using `signer_provider.get_secret_key_for_pubkey()` you can use `signer_provider.derive_contract_signer(offered_contract.keys_id)`

Channels are the only implementation that uses this method. And you don't have to have a conditional for the correct funding pubkey. The funding pubkey comes from deriving a contract signer with the keys id and is stored in `OfferedContract`

matthewjablack replied:
Good catch, fixed in ec6cb89

matthewjablack committed:
refactor: use derive_contract_signer for close

Tibo-lg commented on dlc-messages/src/channel.rs:
```diff
@@ -573,7 +573,6 @@ impl_dlc_writeable!(CollaborativeCloseOffer, {
     derive(serde::Serialize, serde::Deserialize),
     serde(rename_all = "camelCase")
 )]
-
```
Undo

matthewjablack replied:
Undoing causes clippy errors

```
Run cargo clippy --no-default-features --features no-std -p dlc-messages -- -D warnings
warning: /home/runner/work/rust-dlc/rust-dlc/p2pd-oracle-client/Cargo.toml: no edition set: defaulting to the 2015 edition while the latest is 2024
warning: /home/runner/work/rust-dlc/rust-dlc/dlc-trie/Cargo.toml: no edition set: defaulting to the 2015 edition while the latest is 2024
warning: /home/runner/work/rust-dlc/rust-dlc/dlc/Cargo.toml: no edition set: defaulting to the 2015 edition while the latest is 2024
warning: /home/runner/work/rust-dlc/rust-dlc/bitcoin-test-utils/Cargo.toml: no edition set: defaulting to the 2015 edition while the latest is 2024
warning: /home/runner/work/rust-dlc/rust-dlc/dlc-messages/Cargo.toml: no edition set: defaulting to the 2015 edition while the latest is 2024
 Downloading crates ...
  Downloaded lightning-types v0.1.0
  Downloaded bech32 v0.9.1
  Downloaded possiblyrandom v0.2.0
  Downloaded hashbrown v0.13.2
  Downloaded lightning-invoice v0.32.0
  Downloaded libm v0.2.15
  Downloaded lightning v0.0.125
   Compiling secp256k1-sys v0.10.1
   Compiling bitcoin v0.32.7
   Compiling libc v0.2.174
   Compiling libm v0.2.15
    Checking cfg-if v1.0.1
    Checking bech32 v0.9.1
    Checking hashbrown v0.13.2
    Checking getrandom v0.2.16
    Checking possiblyrandom v0.2.0
   Compiling secp256k1-zkp-sys v0.10.1
    Checking secp256k1 v0.29.1
    Checking secp256k1-zkp v0.11.0
    Checking lightning-types v0.1.0
    Checking miniscript v12.3.5
    Checking lightning-invoice v0.32.0
    Checking lightning v0.0.125
    Checking dlc v0.7.1 (/home/runner/work/rust-dlc/rust-dlc/dlc)
    Checking dlc-messages v0.7.1 (/home/runner/work/rust-dlc/rust-dlc/dlc-messages)
error: empty line after outer attribute
   --> dlc-messages/src/channel.rs:571:1
    |
571 | / #[cfg_attr(
572 | |     feature = "use-serde",
573 | |     derive(serde::Serialize, serde::Deserialize),
574 | |     serde(rename_all = "camelCase")
575 | | )]
576 | |
    | |_^
577 |   /// Message used to reject an received offer.
578 |   pub struct Reject {
    |   ----------------- the attribute applies to this struct
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#empty_line_after_outer_attr
    = note: `-D clippy::empty-line-after-outer-attr` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(clippy::empty_line_after_outer_attr)]`
    = help: if the empty line is unintentional, remove it

error: could not compile `dlc-messages` (lib) due to 1 previous error
Error: Process completed with exit code 101.
```

Tibo-lg replied:
Oh ok my bad!

Tibo-lg commented on dlc-messages/src/lib.rs:
```diff
@@ -500,12 +501,57 @@ impl_dlc_writeable!(SignDlc, {
+    /// The payout amount for the accept party in satoshis.
+    pub accept_payout: Amount,
+    /// Serial id for the funding input.
+    pub fund_input_serial_id: u64,
+    /// The funding inputs to use.
+    pub funding_inputs: Vec<FundingInput>,
+    /// The funding signatures.
+    pub funding_signatures: FundingSignatures,
```
I think these fields are never used?

matthewjablack replied:
These fields are necessary to ensure the offeror of the close contract can avoid a free option problem. They could offer a close to the counterparty, and the counterparty does not respond. 

https://github.com/discreetlogcontracts/dlcspecs/issues/161#issuecomment-867819121

matthewjablack replied:
I've added an additional optional argument for `create_collaborative_close_transaction` to include additional inputs in [a915378](https://github.com/p2pderivatives/rust-dlc/pull/255/commits/a9153786317c2270b6376a2cb351c3509849d7a9)

Tibo-lg replied:
Right now the way you set those fields does not make sense, you are setting them to the funding inputs and signatures, which matches with the names but not what you're describing here. If you want to be able to double spend the closing transaction then the inputs and signatures (actually a single one should be enough) should be set to utxos from the wallet. Honestly I feel the best is either you fully implement this (having it optional if you prefer), or just remove it.

Tibo-lg commented on dlc-messages/src/lib.rs:
```diff
@@ -500,12 +501,57 @@ impl_dlc_writeable!(SignDlc, {
+    /// The signature for the closing transaction.
+    pub close_signature: Signature,
+    /// The payout amount for the offer party in satoshis.
+    pub offer_payout: Amount,
```
You don't need to pass that it can be re-computed from `total_collateral`

matthewjablack replied:
Great point, updated in [5e06374](https://github.com/p2pderivatives/rust-dlc/pull/255/commits/5e06374092a8345fe2da40504585860f746b3ba9)

Tibo-lg commented on dlc-manager/src/manager.rs:
```diff
@@ -895,6 +992,86 @@ where
+        Ok((close_message, counter_party))
+    }
+
+    /// Accepts a cooperative close request by verifying the counter party's signature,
```
I don't think you are verifying the signature (not that you need to, since if it's incorrect the broadcast will fail)

matthewjablack replied:
Good catch, updated the comment in [9d84b29](https://github.com/p2pderivatives/rust-dlc/pull/255/commits/9d84b29b372bc2c7486a4127e0c538bbe6294a27)

Tibo-lg commented on dlc-manager/tests/manager_execution_tests.rs:
```diff
@@ -806,93 +848,211 @@ fn manager_execution_test(test_params: TestParams, path: TestPath, manual_close:
+                    assert_contract_state!(alice_manager_send, contract_id, Closed);
+
+                    // Verify the close transaction was properly broadcast and confirmed
+                    let _close_txid = {
```
Remove this block if you don't need this

matthewjablack replied:
Good point, removed in [36c1e23](https://github.com/p2pderivatives/rust-dlc/pull/255/commits/36c1e2394d22d4387cfce220cd5b30d13784ad76)

Tibo-lg commented on dlc-manager/tests/manager_execution_tests.rs:
```diff
@@ -806,93 +848,211 @@ fn manager_execution_test(test_params: TestParams, path: TestPath, manual_close:
 
-                    periodic_check!(second, contract_id, Confirmed);
+                    match path {
+                        TestPath::Close => {
```
I would prefer that you don't refactor the rest of the tests to keep the PR focused, feel free to do another refactor PR.

matthewjablack replied:
Undid test refactor in [36c1e23](https://github.com/p2pderivatives/rust-dlc/pull/255/commits/36c1e2394d22d4387cfce220cd5b30d13784ad76)

Tibo-lg commented on dlc-manager/src/manager.rs:
```diff
@@ -801,9 +770,26 @@ where
             .blockchain
             .get_transaction_confirmations(&broadcasted_txid)?;
         if confirmations >= NB_CONFIRMATIONS {
+            // Check if this is a cooperative close (no attestations) or a CET close (with attestations)
+            let (signed_cet, pnl) = if contract.attestations.is_none() {
+                // Cooperative close - no signed_cet in the final closed contract
+                let pnl = contract
+                    .signed_contract
+                    .accepted_contract
+                    .compute_pnl(&contract.signed_cet)?;
+                (None, pnl)
+            } else {
+                // CET close - include the signed_cet
+                let pnl = contract
+                    .signed_contract
+                    .accepted_contract
+                    .compute_pnl(&contract.signed_cet)?;
+                (Some(contract.signed_cet.clone()), pnl)
+            };
+
```
Why do you differentiate here?

matthewjablack replied:
Good point, no need to differentiate, fixed in [9d84b29](https://github.com/p2pderivatives/rust-dlc/pull/255/commits/9d84b29b372bc2c7486a4127e0c538bbe6294a27)

Tibo-lg commented on dlc-manager/src/manager.rs:
```diff
@@ -827,6 +810,57 @@ where
     }
 
+    /// Check for pending cooperative close transactions
+    fn check_pending_close_transactions(&self) -> Result<(), Error> {
```
This logic should be implemented within `check_confirmed_contracts`, unless there is a reason to have it separate? Right now that means having to go two times through the confirmed contracts.

matthewjablack replied:
Moved to `check_confirmed_contracts` in [9d84b29](https://github.com/p2pderivatives/rust-dlc/pull/255/commits/9d84b29b372bc2c7486a4127e0c538bbe6294a27)

Tibo-lg commented on dlc-manager/src/manager.rs:
```diff
@@ -827,6 +810,57 @@ where
+                    let pnl = contract.accepted_contract.compute_pnl(pending_close_tx)?;
+                    let closed_contract = ClosedContract {
+                        attestations: None, // Cooperative close has no attestations
+                        signed_cet: None,   // Cooperative close doesn't use a CET
```
Why set it to None, I think it's fine to save the closing transaction here, unless it breaks something?

matthewjablack replied:
You're right, signed_cet can be the closing tx, which is must more elegant. Updated in [9d84b29](https://github.com/p2pderivatives/rust-dlc/pull/255/commits/9d84b29b372bc2c7486a4127e0c538bbe6294a27)

matthewjablack committed:
feat: remove offer_payout from CloseDlc message

- rm offer_payout field as it can be computed from total_collateral - accept_payout
- add fee_rate_per_vb field to prevent free option problem
- update serialization accordingly

matthewjablack committed:
refactor: simplify cooperative close logic & improve comments

- rm differentiation between cooperative close and CET close in check_preclosed_contracts
- update accept_cooperative_close comment to be more accurate about what it does
- move pending close transaction logic into check_confirmed_contracts
- save closing transaction instead of setting to None

matthewjablack committed:
feat: add additional inputs support to prevent free option problem

- add additional_inputs parameter to offer_collaborative_close functions
- update create_collaborative_close_transaction to accept additional inputs
- add documentation explaining the free option problem prevention
- keep funding inputs in CloseDlc message to prevent free option problem

matthewjablack committed:
test: update tests for new cooperative close API & behavior

- add test for collaborative close with additional inputs
- update existing tests to use new API with additional_inputs parameter
- add test for free option problem prevention
- update test structure to reflect simplified cooperative close logic

matthewjablack committed:
fix: update test files and clean up code

- update close_msg.json test file to match new CloseDlc structure
- rm offerPayout field and add feeRatePerVb field
- clean up extra blank lines in manager.rs
- fixes test failures caused by structure changes
